r/R/dataset-format.R

```diff
@@ -122,6 +122,31 @@ CsvFileFormat$create <- function(...,
                                  opts = csv_file_format_parse_options(...),
                                  convert_options = csv_file_format_convert_opts(...),
                                  read_options = csv_file_format_read_opts(...)) {
+
+  options <- list(...)
+  schema  <- options[["schema"]]
+
+  column_names     <- read_options$column_names
+  schema_names     <- names(schema)
+
+  if (!is.null(schema) & !identical(schema_names, column_names)) {
+
+    # Element wise comparison and set differnce of column_names and names in schema
+    mismatch_colnames <- column_names[match(column_names, schema_names, nomatch = 0) != seq(column_names)]
+    not_in_schema     <- setdiff(schema_names, column_names)
+
+    abort(c(
+            paste(
+                "column_names not matching or not found in schema-names:",
+                deparse1(c(mismatch_colnames, not_in_schema))
+            ),
+            i = "Set column_names to match names of schema",
+            i = "Omit the column_names argument",
+            i = "Omit the read_options argument"
```

Comment:
I really love that you've come up with helpful solutions for the end-user here.  However, I have a concern that given that there are multiple paths that users can take to get here, and `read_options` isn't an argument that the user is necessarily aware of (it's a few levels down in a function they wouldn't be likely to call directly), and so this error message might be a little confusing - it may be sufficient just to state what's wrong, without any suggestions.

Could I get your thoughts here @jonkeane?

Reply:
I agree with @thisisnic, I really like solutions provided in error messages generally. But here, since there are so many ways one can get to this (and not all of the solutions would work in all of those paths), it's probably better to simply error with the problem up above